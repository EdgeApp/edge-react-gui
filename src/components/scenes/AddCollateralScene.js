// @flow

import s from '../../locales/strings.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { ManageCollateralScene } from './ManageCollateralScene.js'

type Props = {
  navigation: NavigationProp<'addCollateralScene'>,
  route: RouteProp<'addCollateralScene'>
}

export const AddCollateralScene = (props: Props) => {
  const { navigation, route } = props
  const { borrowEngine } = route.params

  return ManageCollateralScene({
    borrowEngine,
    defaultTokenId: borrowEngine.collaterals[0].tokenId,
    action: async req => await borrowEngine.deposit(req),
    actionWallet: 'fromWallet',
    ltvType: 'collaterals',
    ltvChange: 'decrease',

    showExchangeRateTile: true,
    showTotalDebtTile: true,
    showTotalCollateralTile: true,

    headerText: s.strings.loan_add_collateral,
    goBack: () => navigation.pop()
  })
}
