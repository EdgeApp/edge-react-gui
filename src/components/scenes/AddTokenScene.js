// @flow

import * as React from 'react'
import { ActivityIndicator, Alert, ScrollView, StyleSheet, View } from 'react-native'

import { addNewToken } from '../../actions/AddTokenActions.js'
import { MAX_TOKEN_CODE_CHARACTERS } from '../../constants/WalletAndCurrencyConstants.js'
import s from '../../locales/strings.js'
import { PrimaryButton } from '../../modules/UI/components/Buttons/PrimaryButton.ui.js'
import { FormattedText as Text } from '../../modules/UI/components/FormattedText/FormattedText.ui.js'
import { THEME } from '../../theme/variables/airbitz.js'
import { connect } from '../../types/reactRedux.js'
import { type RouteProp } from '../../types/routerTypes.js'
import type { CustomTokenInfo, GuiWallet } from '../../types/types.js'
import { scale } from '../../util/scaling.js'
import { decimalPlacesToDenomination } from '../../util/utils.js'
import { FormField } from '../common/FormField.js'
import { SceneWrapper } from '../common/SceneWrapper.js'

type OwnProps = {
  route: RouteProp<'addToken'>
}

type DispatchProps = {
  addNewToken: (walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, type: string) => void
}

type StateProps = {
  addTokenPending: boolean,
  currentCustomTokens: CustomTokenInfo[],
  wallet: GuiWallet
}

type State = {
  currencyName: string,
  currencyCode: string,
  contractAddress: string,
  decimalPlaces: string,
  enabled?: boolean
}

type Props = OwnProps & StateProps & DispatchProps

class AddToken extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    const { route } = this.props
    const { currencyName = '', currencyCode = '', contractAddress = '', decimalPlaces = '' } = route.params
    this.state = {
      currencyName,
      currencyCode,
      contractAddress,
      decimalPlaces
    }
  }

  render() {
    const { addTokenPending } = this.props
    return (
      <SceneWrapper background="body">
        <ScrollView style={styles.container}>
          <View style={styles.instructionalArea}>
            <Text style={styles.instructionalText}>{s.strings.addtoken_top_instructions}</Text>
          </View>
          <View style={styles.nameArea}>
            <FormField
              value={this.state.currencyName}
              onChangeText={this.onChangeName}
              autoCapitalize="words"
              autoFocus
              label={s.strings.addtoken_name_input_text}
              returnKeyType="done"
              autoCorrect={false}
            />
          </View>
          <View style={styles.currencyCodeArea}>
            <FormField
              value={this.state.currencyCode}
              onChangeText={this.onChangeCurrencyCode}
              autoCapitalize="characters"
              label={s.strings.addtoken_currency_code_input_text}
              returnKeyType="done"
              autoCorrect={false}
              maxLength={MAX_TOKEN_CODE_CHARACTERS}
            />
          </View>
          <View style={styles.contractAddressArea}>
            <FormField
              value={this.state.contractAddress}
              onChangeText={this.onChangeContractAddress}
              label={s.strings.addtoken_contract_address_input_text}
              returnKeyType="done"
              autoCorrect={false}
            />
          </View>
          <View style={styles.decimalPlacesArea}>
            <FormField
              value={this.state.decimalPlaces}
              onChangeText={this.onChangeDecimalPlaces}
              label={s.strings.addtoken_denomination_input_text}
              returnKeyType="done"
              autoCorrect={false}
              keyboardType="numeric"
            />
          </View>
          <View style={styles.buttonsArea}>
            <PrimaryButton style={styles.saveButton} onPress={this._onSave}>
              {addTokenPending ? <ActivityIndicator color={THEME.COLORS.ACCENT_MINT} /> : <PrimaryButton.Text>{s.strings.string_save}</PrimaryButton.Text>}
            </PrimaryButton>
          </View>
          <View style={styles.bottomPaddingForKeyboard} />
        </ScrollView>
      </SceneWrapper>
    )
  }

  onChangeName = (input: string) => {
    this.setState({
      currencyName: input
    })
  }

  onChangeCurrencyCode = (input: string) => {
    this.setState({
      currencyCode: input
    })
  }

  onChangeDecimalPlaces = (input: string) => {
    this.setState({
      decimalPlaces: input
    })
  }

  onChangeContractAddress = (input: string) => {
    this.setState({
      contractAddress: input.trim()
    })
  }

  _onSave = () => {
    const currencyCode = this.state.currencyCode.toUpperCase()
    this.setState(
      {
        currencyCode
      },
      () => {
        const { route, currentCustomTokens, wallet, addNewToken } = this.props
        const { currencyName, decimalPlaces, contractAddress } = this.state
        const { walletId } = route.params

        const currentCustomTokenIndex = currentCustomTokens.findIndex(item => item.currencyCode === currencyCode)
        const metaTokensIndex = wallet.metaTokens.findIndex(item => item.currencyCode === currencyCode)
        // if token is hard-coded into wallets of this type
        if (metaTokensIndex >= 0) Alert.alert(s.strings.manage_tokens_duplicate_currency_code)
        // if that token already exists and is visible (ie not deleted)
        if (currentCustomTokenIndex >= 0 && currentCustomTokens[currentCustomTokenIndex].isVisible !== false) {
          Alert.alert(s.strings.manage_tokens_duplicate_currency_code)
        } else {
          if (currencyName && currencyCode && decimalPlaces && contractAddress) {
            const denomination = decimalPlacesToDenomination(decimalPlaces)
            addNewToken(walletId, currencyName, currencyCode, contractAddress, denomination, wallet.type)
          } else {
            Alert.alert(s.strings.addtoken_invalid_information)
          }
        }
      }
    )
  }
}

const rawStyles = {
  container: {
    flex: 1,
    paddingHorizontal: scale(20)
  },

  instructionalArea: {
    paddingVertical: scale(16),
    paddingHorizontal: scale(20)
  },
  instructionalText: {
    fontSize: scale(16),
    textAlign: 'center'
  },

  nameArea: {
    height: scale(70)
  },
  currencyCodeArea: {
    height: scale(70)
  },
  contractAddressArea: {
    height: scale(70)
  },
  decimalPlacesArea: {
    height: scale(70)
  },
  buttonsArea: {
    marginTop: scale(16),
    height: scale(52),
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    alignSelf: 'flex-end',
    paddingVertical: scale(4)
  },
  saveButton: {
    flex: 1,
    marginLeft: scale(2),
    backgroundColor: THEME.COLORS.SECONDARY,
    borderRadius: 3
  },
  bottomPaddingForKeyboard: {
    height: scale(300)
  }
}
const styles: typeof rawStyles = StyleSheet.create(rawStyles)

export const AddTokenScene = connect<StateProps, DispatchProps, OwnProps>(
  (state, { route: { params } }) => ({
    addTokenPending: state.ui.wallets.addTokenPending,
    currentCustomTokens: state.ui.settings.customTokens,
    wallet: state.ui.wallets.byId[params.walletId]
  }),
  dispatch => ({
    addNewToken(walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, walletType: string) {
      dispatch(addNewToken(walletId, currencyName, currencyCode, contractAddress, denomination, walletType))
    }
  })
)(AddToken)
