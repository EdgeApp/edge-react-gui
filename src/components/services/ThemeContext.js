// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
