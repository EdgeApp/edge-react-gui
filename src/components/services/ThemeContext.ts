import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
