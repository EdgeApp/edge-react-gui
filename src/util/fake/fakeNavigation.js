// @flow

import { type NavigationProp } from '../../types/routerTypes.js'

export const fakeNavigation: NavigationProp<any> = {
  addListener(event, callback) {
    return () => undefined
  },
  isFocused() {
    return true
  },

  navigate(name, params) {},
  push(name, params) {},
  replace(name, params) {},
  setParams(params) {},
  jumpTo(name, params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  closeDrawer() {},
  openDrawer() {},
  toggleDrawer() {},

  state: {},
  currentScene: {}
}
