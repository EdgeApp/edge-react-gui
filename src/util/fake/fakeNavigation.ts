import { NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener(event, callback) {
    return () => undefined
  },
  isFocused() {
    return true
  },

  navigate(name, params) {},
  push(name, params) {},
  replace(name, params) {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  closeDrawer() {},
  openDrawer() {},
  toggleDrawer() {},

  state: {}
}
