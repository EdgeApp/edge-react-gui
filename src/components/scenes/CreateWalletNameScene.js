// @flow

import * as React from 'react'
import { Alert } from 'react-native'
import { sprintf } from 'sprintf-js'

import { getSpecialCurrencyInfo } from '../../constants/WalletAndCurrencyConstants.js'
import s from '../../locales/strings.js'
import { type TestProps } from '../../types/reactRedux.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { SceneWrapper } from '../common/SceneWrapper'
import { type ThemeProps, withTheme } from '../services/ThemeContext.js'
import { MainButton } from '../themed/MainButton.js'
import { OutlinedTextInput } from '../themed/OutlinedTextInput.js'
import { SceneHeader } from '../themed/SceneHeader'

export type CreateWalletNameOwnProps = {
  navigation: NavigationProp<'createWalletName'>,
  route: RouteProp<'createWalletName'>
}
type Props = CreateWalletNameOwnProps & ThemeProps & TestProps
type State = {
  walletName: string
}

class CreateWalletNameComponent extends React.Component<Props & TestProps, State> {
  constructor(props: Props) {
    super(props)
    const { route } = props
    const { currencyCode, currencyName } = route.params.selectedWalletType

    const name = currencyCode.toLowerCase() === 'xrp' ? 'XRP' : currencyName
    const walletName = sprintf(s.strings.my_crypto_wallet_name, name)
    this.state = { walletName }
  }

  isValidWalletName = () => {
    const { walletName } = this.state
    const isValid = walletName.length > 0

    return isValid
  }

  onNext = () => {
    const { navigation, route } = this.props
    const { cleanedPrivateKey, selectedFiat, selectedWalletType } = route.params
    const { walletType } = selectedWalletType
    const specialCurrencyInfo = getSpecialCurrencyInfo(walletType)

    if (this.isValidWalletName() && specialCurrencyInfo.skipAccountNameValidation && !cleanedPrivateKey) {
      navigation.navigate('createWalletAccountSelect', {
        selectedFiat: selectedFiat,
        selectedWalletType,
        accountName: this.state.walletName,
        existingWalletId: ''
      })
    } else if (this.isValidWalletName()) {
      navigation.navigate('createWalletReview', {
        walletName: this.state.walletName,
        selectedFiat: selectedFiat,
        selectedWalletType: selectedWalletType,
        cleanedPrivateKey
      })
    } else {
      Alert.alert(s.strings.create_wallet_invalid_name, s.strings.create_wallet_enter_valid_name)
    }
  }

  handleChangeWalletName = (walletName: string) => {
    this.setState({ walletName })
  }

  render() {
    const buttonType = this.props.theme.preferPrimaryButton ? 'primary' : 'secondary'

    return (
      <SceneWrapper avoidKeyboard background="theme">
        <SceneHeader withTopMargin title={s.strings.title_create_wallet} />

        <OutlinedTextInput
          onChangeText={this.handleChangeWalletName}
          value={this.state.walletName}
          onSubmitEditing={this.onNext}
          autoCorrect={false}
          returnKeyType="next"
          label={s.strings.fragment_wallets_addwallet_name_hint}
          marginRem={[0, 1.75]}
        />
        <MainButton
          alignSelf="center"
          label={s.strings.string_next_capitalized}
          marginRem={[3, 1]}
          type={buttonType}
          onPress={this.onNext}
          ref={this.props.generateTestHook('CreateWalletNameScene.NextButton')}
        />
      </SceneWrapper>
    )
  }
}

export const CreateWalletName = withTheme(CreateWalletNameComponent)
