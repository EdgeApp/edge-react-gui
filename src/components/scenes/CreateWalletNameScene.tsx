import * as React from 'react'
import { Alert } from 'react-native'
import { sprintf } from 'sprintf-js'

import { getSpecialCurrencyInfo } from '../../constants/WalletAndCurrencyConstants'
import s from '../../locales/strings'
import { NavigationProp, RouteProp } from '../../types/routerTypes'
import { SceneWrapper } from '../common/SceneWrapper'
import { ThemeProps, withTheme } from '../services/ThemeContext'
import { MainButton } from '../themed/MainButton'
import { OutlinedTextInput } from '../themed/OutlinedTextInput'
import { SceneHeader } from '../themed/SceneHeader'

export type CreateWalletNameOwnProps = {
  navigation: NavigationProp<'createWalletName'>
  route: RouteProp<'createWalletName'>
}
type Props = CreateWalletNameOwnProps & ThemeProps
type State = {
  walletName: string
}

class CreateWalletNameComponent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    const { route } = props
    const { currencyCode, currencyName } = route.params.selectedWalletType

    const name = currencyCode.toLowerCase() === 'xrp' ? 'XRP' : currencyName
    const walletName = sprintf(s.strings.my_crypto_wallet_name, name)
    this.state = { walletName }
  }

  isValidWalletName = () => {
    const { walletName } = this.state
    const isValid = walletName.length > 0

    return isValid
  }

  onNext = () => {
    const { navigation, route } = this.props
    const { cleanedPrivateKey, selectedFiat, selectedWalletType } = route.params
    const { walletType } = selectedWalletType
    const specialCurrencyInfo = getSpecialCurrencyInfo(walletType)

    if (this.isValidWalletName() && specialCurrencyInfo.skipAccountNameValidation && !cleanedPrivateKey) {
      navigation.navigate('createWalletAccountSelect', {
        selectedFiat: selectedFiat,
        selectedWalletType,
        accountName: this.state.walletName,
        existingWalletId: ''
      })
    } else if (this.isValidWalletName()) {
      navigation.navigate('createWalletReview', {
        walletName: this.state.walletName,
        selectedFiat: selectedFiat,
        selectedWalletType: selectedWalletType,
        cleanedPrivateKey
      })
    } else {
      Alert.alert(s.strings.create_wallet_invalid_name, s.strings.create_wallet_enter_valid_name)
    }
  }

  handleChangeWalletName = (walletName: string) => {
    this.setState({ walletName })
  }

  render() {
    const buttonType = this.props.theme.preferPrimaryButton ? 'primary' : 'secondary'

    return (
      <SceneWrapper avoidKeyboard background="theme">
        <SceneHeader withTopMargin title={s.strings.title_create_wallet} />

        <OutlinedTextInput
          onChangeText={this.handleChangeWalletName}
          value={this.state.walletName}
          onSubmitEditing={this.onNext}
          autoCorrect={false}
          returnKeyType="next"
          label={s.strings.fragment_wallets_addwallet_name_hint}
          marginRem={[0, 1.75]}
        />
        <MainButton alignSelf="center" label={s.strings.string_next_capitalized} marginRem={[3, 1]} type={buttonType} onPress={this.onNext} />
      </SceneWrapper>
    )
  }
}

export const CreateWalletName = withTheme(CreateWalletNameComponent)
