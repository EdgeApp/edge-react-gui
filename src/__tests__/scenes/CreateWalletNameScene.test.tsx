/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { CreateWalletName } from '../../components/scenes/CreateWalletNameScene'
import { getTheme } from '../../components/services/ThemeContext'
import { fakeNavigation } from '../../util/fake/fakeNavigation'

describe('createWalletName', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      navigation: fakeNavigation,
      route: {
        name: 'createWalletName',
        params: {
          cleanedPrivateKey: 'bitcoinbitcoinbitcoin',
          selectedWalletType: {
            currencyName: 'bitcoin',
            walletType: 'wallet:bitcoin',
            symbolImage: 'BTC',
            currencyCode: 'BTC'
          },
          selectedFiat: {
            label: 'USD',
            value: 'USD'
          }
        }
      },
      theme: getTheme()
    }
    const actual = renderer.render(<CreateWalletName {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
