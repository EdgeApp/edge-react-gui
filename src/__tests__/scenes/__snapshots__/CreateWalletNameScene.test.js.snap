// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createWalletName should render with loading props 1`] = `
<SceneWrapper
  avoidKeyboard={true}
  background="theme"
>
  <WithTheme(SceneHeaderComponent)
    title="Create Wallet"
    withTopMargin={true}
  />
  <ForwardRef
    autoCorrect={false}
    label="New Wallet Name"
    marginRem={
      Array [
        0,
        1.75,
      ]
    }
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    returnKeyType="next"
    value="My bitcoin"
  />
  <ForwardRef
    alignSelf="center"
    label="Next"
    marginRem={
      Array [
        3,
        1,
      ]
    }
    onPress={[Function]}
    type="secondary"
  />
</SceneWrapper>
`;
